src/protocol/sort/shuffle.rs

```diff
@@ -0,0 +1,280 @@
+    /// drawing an element from the hat until no elements remain. The algorithm produces an unbiased
+    /// permutation: every permutation is equally likely. This algorithm takes time proportional to
+    /// the number of items being shuffled and shuffles them in place.
+    #[allow(clippy::cast_possible_truncation)]
+    fn generate_random_sequence(
+        batchsize: usize,
+        index: u64,
+        left_permute: bool,
+        prss: &PrssSpace,
+    ) -> Vec<usize> {
+        let mut permutation: Vec<usize> = (0..batchsize).collect();
+        let rand = prss.generate_values(index.into());
+        let mut rng = if left_permute {
+            StepRng::new(rand.0 as u64, 1)
+        } else {
+            StepRng::new(rand.1 as u64, 1)
+        };
+        permutation.shuffle(&mut rng);
+        permutation
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    fn generate_random_sequences(
+        &self,
+        left_permute: bool,
+        prss: &PrssSpace,
+        seed: u128,
+    ) -> Vec<usize> {
+        let batchsize = self.input.len();
+        let random_values = prss.generate_values(seed);
+        if left_permute {
+            Self::generate_random_sequence(batchsize, random_values.0 as u64, true, prss)
+        } else {
+            Self::generate_random_sequence(batchsize, random_values.1 as u64, false, prss)
+        }
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    async fn reshare_all_shares<M, G>(
+        &self,
+        ctx: &ProtocolContext<'_, G, S>,
+        to_helper: Identity,
+        step: S,
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, S>,
+    {
+        let reshares = self
+            .input
+            .iter()
+            .enumerate()
+            .map(|(index, input)| async move {
+                Reshare::new(*input)
+                    .execute(ctx, RecordId::from(index as u32), step, to_helper)
+                    .await
+            });
+        try_join_all(reshares).await
+    }
+
+    /// `single_shuffle` is called for the helpers
+    /// i) 2/3 helpers generate random sequence using shared prss random number
+    /// ii) 2/3 helpers apply the permutation to their shares
+    /// iii) reshare to `to_helper`
+    async fn single_shuffle<M: Mesh, G: Gateway<M, S>>(
+        &mut self,
+        ctx: &ProtocolContext<'_, G, S>,
+        seed: u128,
+        to_helper: Identity,
+        which_step: ShuffleStep,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let step = (self.step_fn)(which_step);
+        let prss = &ctx.participant[step];
+        let channel = ctx.gateway.get_channel(step);
+
+        if to_helper != channel.identity() {
+            let mut permute = if to_helper.peer(Direction::Left) == channel.identity() {
+                self.generate_random_sequences(true, prss, seed)
+            } else {
+                self.generate_random_sequences(false, prss, seed)
```

Comment:
I do not like using a boolean for this first argument. Can we use either an enum value, or make two different functions? From the caller's perspective, reading this code it's not clear what this bool means. I have to consult the function definition.

Reply:
Changed this to use Direction::Left and Direction::Right